solutions/rust/05-echo/code/src/resp.rs

```diff
@@ -0,0 +1,209 @@
+                } else {
+                    return Err(Error::msg("connection closed unexpectedly"));
+                }
+            }
+
+            if let Some((value, _)) = parse_message(self.buffer.split())? {
+                return Ok(Some(value));
+            }
+        }
+    }
+
+    pub async fn write_value(&mut self, value: Value) -> Result<()> {
+        self.stream.write(value.encode().as_bytes()).await?;
+
+        Ok(())
+    }
+}
+
+fn parse_message(buffer: BytesMut) -> Result<Option<(Value, usize)>> {
+    match buffer[0] as char {
+        '+' => decode_simple_string(buffer),
+        '*' => decode_array(buffer),
+        '$' => decode_bulk_string(buffer),
+        _ => {
+            return Err(Error::msg("unrecognised message type"));
+        }
+    }
+}
+
+fn decode_simple_string(buffer: BytesMut) -> Result<Option<(Value, usize)>> {
+    if let Some((line, len)) = get_line(&buffer[1..]) {
+        let str = parse_string(line)?;
+
+        Ok(Some((Value::String(str), len + 1)))
+    } else {
+        Ok(None)
+    }
+}
+
+fn decode_array(buffer: BytesMut) -> Result<Option<(Value, usize)>> {
+    let (array_length, mut bytes_consumed) = if let Some((line, len)) = get_line(&buffer[1..]) {
+        let array_length = parse_integer(line)?;
+
+        (array_length, len + 1)
+    } else {
+        return Ok(None);
+    };
+
+    let mut items: Vec<Value> = Vec::new();
+    for _ in 0..array_length {
+        if let Some((v, len)) = parse_message(BytesMut::from(&buffer[bytes_consumed..]))? {
+            items.push(v);
+            bytes_consumed += len
+        } else {
+            return Ok(None);
+        }
+    }
+
+    return Ok(Some((Value::Array(items), bytes_consumed)));
+}
+
+fn decode_bulk_string(buffer: BytesMut) -> Result<Option<(Value, usize)>> {
+    let (bulk_length, bytes_consumed) = if let Some((line, len)) = get_line(&buffer[1..]) {
+        let bulk_length = parse_integer(line)?;
+
+        (bulk_length, len + 1)
+    } else {
+        return Ok(None);
+    };
+
+    let end_of_bulk = bytes_consumed + (bulk_length as usize);
+    let end_of_bulk_line = end_of_bulk + 2;
+
+    return if end_of_bulk_line <= buffer.len() {
+        Ok(Some((
+            Value::Bulk(parse_string(&buffer[bytes_consumed..end_of_bulk])?),
+            end_of_bulk_line,
+        )))
+    } else {
+        Ok(None)
+    };
+}
+
+fn get_line(buffer: &[u8]) -> Option<(&[u8], usize)> {
+    for i in 1..buffer.len() {
+        if buffer[i - 1] == CARRIAGE_RETURN && buffer[i] == NEWLINE {
+            return Some((&buffer[0..(i - 1)], i + 1));
+        }
+    }
+
+    return None;
+}
+
+fn parse_string(bytes: &[u8]) -> Result<String> {
+    String::from_utf8(bytes.to_vec()).map_err(|_| Error::msg("Could not parse string"))
+}
+
+fn parse_integer(bytes: &[u8]) -> Result<i64> {
+    let str_integer = parse_string(bytes)?;
+    (str_integer.parse::<i64>()).map_err(|_| Error::msg("Could not parse integer"))
+}
+
+#[cfg(test)]
```

Comment:
@ThetaSinner let's remove these tests - we intentionally don't support tests since they're extra work to maintain. Not sure if we had tests in the previous stages - if we did, let's please remove them too. 

We want to support tests within solutions in the long term, but it'll require some proper thinking to ensure the user experience is ideal & we're able to maintain these effectively.